runtime/parachains/src/paras_inherent.rs

```diff
@@ -290,239 +488,1609 @@ pub mod pallet {
-fn limit_backed_candidates<T: Config>(
+/// `candidate_has_concluded_invalid_dispute` must return `true` if the candidate
+/// is disputed, false otherwise
+fn sanitize_backed_candidates<T: crate::inclusion::Config, F: Fn(CandidateHash) -> bool>(
+	relay_parent: T::Hash,
 	mut backed_candidates: Vec<BackedCandidate<T::Hash>>,
+	candidate_has_concluded_invalid_dispute: F,
+	scheduled: &[CoreAssignment],
 ) -> Vec<BackedCandidate<T::Hash>> {
-	const MAX_CODE_UPGRADES: usize = 1;
+	// Remove any candidates that were concluded invalid.
+	backed_candidates.retain(|backed_candidate| {
+		!candidate_has_concluded_invalid_dispute(backed_candidate.candidate.hash())
+	});
+
+	// Assure the backed candidate's `ParaId`'s core is free.
+	// This holds under the assumption that `Scheduler::schedule` is called _before_.
+	// Also checks the candidate references the correct relay parent.
+	let scheduled_paras_set = scheduled
+		.into_iter()
+		.map(|core_assignment| core_assignment.para_id)
+		.collect::<BTreeSet<_>>();
+	backed_candidates.retain(|backed_candidate| {
+		let desc = backed_candidate.descriptor();
+		desc.relay_parent == relay_parent && scheduled_paras_set.contains(&desc.para_id)
+	});
+
+	backed_candidates
+}
 
-	// Ignore any candidates beyond one that contain code upgrades.
-	//
-	// This is an artificial limitation that does not appear in the guide as it is a practical
-	// concern around execution.
-	{
-		let mut code_upgrades = 0;
-		backed_candidates.retain(|c| {
-			if c.candidate.commitments.new_validation_code.is_some() {
-				if code_upgrades >= MAX_CODE_UPGRADES {
-					return false
-				}
+fn compute_entropy<T: Config>(parent_hash: T::Hash) -> [u8; 32] {
+	const CANDIDATE_SEED_SUBJECT: [u8; 32] = *b"candidate-seed-selection-subject";
+	let vrf_random = CurrentBlockRandomness::<T>::random(&CANDIDATE_SEED_SUBJECT[..]).0;
+	let mut entropy: [u8; 32] = CANDIDATE_SEED_SUBJECT.clone();
+	if let Some(vrf_random) = vrf_random {
+		entropy.as_mut().copy_from_slice(vrf_random.as_ref());
+	} else {
+		// in case there is no vrf randomness present, we utilize the relay parent
+		// as seed, it's better than a static value.
+		log::warn!(target: LOG_TARGET, "CurrentBlockRandomness did not provide entropy");
+		entropy.as_mut().copy_from_slice(parent_hash.as_ref());
+	}
+	entropy
+}
 
-				code_upgrades += 1;
+/// Limit disputes in place.
+///
+/// Returns the unused weight of `remaining_weight`.
+fn limit_disputes<T: Config>(
+	disputes: &mut MultiDisputeStatementSet,
+	remaining_weight: Weight,
+	rng: &mut rand_chacha::ChaChaRng,
+) -> Weight {
+	let mut remaining_weight = remaining_weight;
+	let disputes_weight = dispute_statements_weight::<T>(&disputes);
+	if disputes_weight > remaining_weight {
+		// Sort the dispute statements according to the following prioritization:
+		//  1. Prioritize local disputes over remote disputes.
+		//  2. Prioritize older disputes over newer disputes.
+		disputes.sort_by(|a, b| {
```

Comment:
Looks like we can use `sort_unstable_by` here (https://doc.rust-lang.org/std/vec/struct.Vec.html#method.sort_unstable_by), which is faster according to the docs

Reply:
no please: https://github.com/paritytech/substrate/pull/6754

Reply:
Ok @shawntabrizi pointed out this is not worth the undefined behavior and can lead to consensus issues https://github.com/paritytech/substrate/pull/6754